Pidgin/Parser.Number.cs

```diff
@@ -143,5 +145,54 @@ private static long GetLetterOrDigitValueLong(char c)
         }
         private static long GetUpperLetterOffsetLong(char c) => c - 'A';
         private static long GetLowerLetterOffsetLong(char c) => c - 'a';
+		
+		/// <summary>
+        /// A parser which parses a floating point number with an optional sign.
+        /// The resulting <c>double</c> is not checked for overflow.
+        /// </summary>
+        /// <returns>A parser which parses a floating point number with an optional sign</returns>
+        public static Parser<char, double> Float { get; } = RealNum();
+
+        /// <summary>
+        /// A parser which parses a floating point number with an optional sign.
+        /// The resulting <c>double</c> is not checked for overflow.
+        /// </summary>
+        /// <returns>A parser which parses a floating point number with an optional sign</returns>
+        public static Parser<char, double> RealNum()
+            => Map(
+                (sign, num) => sign.HasValue && sign.Value == '-' ? -num : num,
+                (Char('-').Or(Char('+'))).Optional(),
+                UnsignedReal()
+            ).Labelled($"real number");
+			
+		/// <summary>
+		/// A parser which parses a floating point number without a sign.
+		/// The resulting <c>double</c> is not checked for overflow.
+		/// </summary>
+		/// <returns>A parser which parses a floating point number without a sign.</returns>
+
+		public static Parser<char, double> UnsignedReal()
+			=> Map(
+				(integerPart, _, decimalPart) => integerPart.HasValue ? integerPart.Value + decimalPart : decimalPart,
+				UnsignedInt(10).Optional(),
+				Char('.'),
+				DigitCharDouble().ChainAtLeastOnceL(
+					() => 
+						{
+							var sb = new PooledStringBuilder();
+							sb.Append('.');
+							return sb;
+						},
+					(sb, c) =>
+						{ sb.Append((char) c); return sb; }
+					).Select(sb =>
+						{ return Convert.ToDouble(sb.GetStringAndClear()); }
```

Comment:
`GetStringAndClear` unfortunately allocates a string (so that it can return it). We know that we're not keeping the resulting string around so it's a waste of memory to allocate it. I think a better bet here would be to use a different type of string builder which unsafely mutates a pooled string in place. Untested sketch:

```csharp
internal struct InPlacePooledStringBuilder
{
    private string _buffer;
    private int _length;

    public string Get() => _buffer;

    public unsafe void Add(char c)
    {
        if (_buffer == null)
        {
            _buffer = GetFromPool(16);
        }
        if (_length == _buffer.Length)
        {
            var newBuffer = GetFromPool(_length * 2);
            fixed (char* p = newString)
            {
                for (var i = 0; i < _length; i++)
                {
                    (*p + i) = _buffer[i];
                }
            }
            ReturnToPool(_buffer);
            _buffer = newBuffer;
        }
        fixed (char* p = _buffer)
        {
            *(p + _length) = c;
        }
        _length++;
    }

    public void Dispose()
    {
        ReturnToPool(_buffer);
    }

    private static string GetFromPool(int minLength) { /* todo */ }
    private static void ReturnToPool(string buffer) { /* todo */ }
}
```

You'd need to make sure to null out the string's suffix at some point - either null out the whole thing when you get it from the pool, or just the part which wasn't written to when you call `Get`. Also need to double check that a string containing a float _followed by nulls_ is accepted by `double.Parse`.

I'm envisioning the pool would consist of a collection of buckets, whose contents are strings with a minimum length of 16, 32, 64, etc. It'd need to be thread safe, can probably use a spin-lock for that.

Reply:
As a long term Windows developer, I am very familiar with the concept you describe and agree that it adds a performance edge. You have certainly done a great job making Pidgin as fast as it is! so its worth considering.

My biggest concern is use on non-Windows systems. I am less familiar with the 'under the hood' differences between the CLR run times between Windows and Linux.  In Windows strings are stored as COM BSTR structures that include a count (string address -4 (32bits max length)), and most (but not all!) Windows functions rely on that count rather than null termination.  Regardless we would have to look at the .NET conversion routines to see if the are length based.  Of course nulling the buffer on allocation revolves that to a certain extent.

This may be a non-issue and I defer to your feedback.  

In this case there is a single allocation after we have determined that it is a floating point number so the impact will be smaller.  Obviously when it is part of a larger parse structure it may be repeated multiple times.

Reply:
Sorry, I should have clarified - I'm not suggesting nulling out the string because .NET strings are null-terminated (they're not - any string can contain any number of nulls at any location). It needs to be nulled out to erase data from previous usages of the pooled string.

Reply:
Ok, If I understand you correctly ...

GetFromPool() would get from a pool of pre-allocated, fixed length strings. 

Updating or adding to this string would be done with a 'map' like function that would assume a wide character width (16 bits) and writing the data directly into the buffer at an incrementing offset. The string would be initialized (or logically terminated) with a space(s) or nulls, which should stop the ultimate conversion.  By using a fixed length string, you are not affecting the  BSTR structure (or any other underlying string data).

We would then use the original string pointer from when it was allocated and call the .NET Convert.ToDouble() to create the converted value without having to (re) allocate the string. Destroying the pointer before it goes out of scope so that the original string is not deallocated.

Sorry if I am be obtuse! 

Assuming my understanding is correct ... It does not appear that the .NET convert method will work with a string that does not have a reasonable structure ...

Quick test:
x = Convert.ToDouble("123.456");   <- works
x = Convert.ToDouble("123.456 X"); <- fails (with extra text)  
x = Convert.ToDouble("123.456\0X"); <- fails (with extra text)  

same as ...
s = "123.456 X";
try{ x = Convert.ToDouble(s);} catch {};
s = "123.456" + Convert.ToChar(0x0).ToString() + "X";
try{ x = Convert.ToDouble(s);} catch {};

I am not sure how we can get from a fixed length string back into something that the convert method would like (without creating a new string).

Reply:
Thanks, I confirm that is the case.

Rather than create a new pooling technology, what do you think about using Span?  I spent a lot of time yesterday looking at it and was very impressed. The fact that the Double.Parse() command takes a span as a native overload is very helpful -- no conversions required. It is supported in NETCOREAPP2_1 and greater.

public static Parser<char, double> UnsignedReal()
            => Map(
                (integerPart, _, decimalPart) => integerPart.HasValue ? integerPart.Value + decimalPart : decimalPart,
                UnsignedInt(10).Optional(),
                Char('.'),
                DigitCharDouble().ChainAtLeastOnceL(
                    () =>
                    {
                        var sb = new PooledStringBuilder();
                        sb.Append('.');
                        return sb;
                    },
                    (sb, c) =>
                    { sb.Append(c); return sb; }
                    ).Select(sb =>
                    {
                           #if <some build condition>
                           ReadOnlySpan<char> csp = sb.Buffer;
                          var x = Double.Parse(csp, System.Globalization.NumberStyles.Float);
                          sb.Clear();
                          return x;
                          #else
                          return Convert.ToDouble(sb.GetStringAndClear());
                          #endif
                     }

// Approximate changes to PooledStringBuilder
 public char[] Buffer()
        {
            get  
            {
                // Null remainder of buffer on access
                for (int i = _length; i < _buffer.Length; i++)
                {
                    _buffer[i] = '\0';
                }
                return _buffer;
            }
        }

public void Clear()
        {
            ArrayPool<char>.Shared.Return(_buffer);
            _buffer = null;
            _length = 0;
        }

We could conditionally compile using sb,GetStringAndClear(); and the ReadOnlySpan technology. The latest platform performing the best.

I verified the general concept would work in our main project which is 2.1.  That being said,
I tried to get this working in the current Pidgin tree with very limited success.  I am a bit stymied by multi-version/multi-project features of this solution and could not get that code to actually compile the Double.Parse() ... simply because it would not recognize the NETCOREAPP2_1 run time environment.  I leave that to you as the expert.

Anyway, I highly recommend this approach.

Reply:
`double.Parse(Span<char>)` is only available in .NET Core, this is a .NET Standard library. I'd rather have it work the same way across all platforms if possible

Reply:
I totally get your point.  We choose .NetCore (in spite of the early pains) because of the performance footprint and ability to more easily run on different platforms.  But as a longtime Microsoft developer, I am keenly aware of legacy!

It was my understanding that the MemoryExtensions Class was part of NETStandard.Library 2.0 (although I havn't been able to confirm this) and that specifically SPAN support is available in NETStandard.Library 2.1.
https://blogs.msdn.microsoft.com/dotnet/2018/11/05/announcing-net-standard-2-1/

That would open it up to the latest and greatest .NET Frameworks.  

What I am suggesting is a short term solution of conditional compile (get string versus span), that would be broadened as time went on. SPAN appears to be such a good technology, that creating an 'unsafe' equivalent would be a short term solution only.

As this Parser is a new feature, the impact would be minimal.

Your call of course.

Reply:
`Span` itself is available on the .NET Standard (I think it's a slightly slower implementation?) but the various library methods like `double.Parse(Span<char>)` or `stream.Read(Span<char>)` are limited to .NET Core. As far as I know there are no plans to roll out those methods to the desktop framework.

Honestly [I am kind of pissed off](https://www.benjamin.pizza/posts/2018-12-06-zooming-in-on-field-accessors.html) about this decision by MS. What's the point in adding all these features designed for library authors if you're not going to support them properly? It'd be preferable if they just directly said they were sunsetting the desktop framework, then library authors could reasonably start dropping support for it.

I would love to use `Span` more extensively throughout the library. It would allow me to delete a lot of code and complexity. For now compat means I can't though.